File: dvd_DgrafX/Source/DiscVisionDeluxeUE/DfisX/Dgyro.cpp
#include "DfisX.hpp"
#include "Daero.hpp"
#include "Dgyro.hpp"

#include "dvd_maths.hpp"

#include <iostream> 
#include <math.h>


namespace DfisX
{
  void step_Dgyro(Throw_Container *throw_container, const float dt)
  {
    // old
    //#moment of inertia      %%%%0.9 is approx k value
    //double Im = mass * (0.9*disc.radius)**2  
    //double Im = d_object.mass * (0.9*0.105)*(0.9*0.105);
    ///below is set as a constant because k and disc.radius are amalgamated into Disc_Model properties
    //double Im = d_object.mass * 0.00893025;

    //new
    // Inertia of a thin disc:
    // Iz =      1/2 * m * r^2
    // Ix = Iy = 1/4 * m * r^2
    const double Ix = 1.0/4.0 * d_object.mass * (d_object.radius*d_object.radius);
    const double Iz = 1.0/2.0 * d_object.mass * (d_object.radius*d_object.radius);

    // ######Gyroscopic Precession
    //    #Wp  is rads/sec of off axis rotation (rolling)
    //    Wp = Pitching moment / (Moment of intertia * angular velocity (of gyro))
    double Wp = 0.0;
    d_forces.gyro_torque_y = 0.0;
    if(abs(d_state.disc_rotation_vel) > 0.1)
    {
        Wp = (d_forces.lift_induced_pitching_moment + d_forces.aero_torque_x) / (Iz * d_state.disc_rotation_vel);
        // compute resulting angular torque from applied pitching moment vel
        const double Wp_d  = (Wp - d_state.disc_pitching_vel) / MAX(dt, CLOSE_TO_ZERO);
        d_forces.gyro_torque_y = Wp_d * Ix;
    }    

    // does this ever get applied?
    //d_forces.gyro_torque_x = d_forces.lift_induced_pitching_moment;

    // Mike: what do we need this for? is this actually the roll change??
    d_statistics.disc_cumulative_roll += RAD_TO_DEG(Wp) * dt;

    //always do this after affecting disc_orient_z_vect
    make_unit_vector(d_state.disc_orient_z_vect);
  }
}

File: Assets/DfisX/Runtime/Core/Dgyro.cs
using Unity.Mathematics;
using static Unity.Mathematics.math;
using float3 = Unity.Mathematics.float3;
using float3x3 = Unity.Mathematics.float3x3;

// ============================================================================
//  Dgyro.cs
//  Port of Dgyro.cpp — gyroscopic precession for disc flight.
//
//  Porting notes:
//  - double → float  (consistent with Daero.cs decision)
//  - The `mode` bool parameter in C++ only enables a block that is entirely
//    commented-out dead code. It is dropped here.
//  - make_unit_vector(disc_orient_z_vect) at the end of C++ Dgyro is omitted:
//    Dpropagate re-normalises all orientation vectors after propagation,
//    making this redundant. The normalise was a safety measure in C++ for
//    the case where Dgyro ran without Dpropagate following it.
//  - friction_input.ang_torque_XYZ and collision_torque_xyz are included in
//    the net moment computation, matching the C++ exactly.
// ============================================================================

namespace DfisX
{
    public static class Dgyro
    {
        const float CLOSE_TO_ZERO = 0.000001f;

        /// <summary>
        /// Compute gyroscopic precession torques for one simulation step.
        /// Writes gyroTorqueX and gyroTorqueY into forcesState.
        ///
        /// Call order inside step_simulation():
        ///   1. StepDaero   (Daero.cs)
        ///   2. StepDgyro   ← this method
        ///   3. Propagate   (Dpropagate.cs)
        /// </summary>
        public static void StepDgyro(ThrowContainer tc, float dt)
        {
            ref DiscState     dState  = ref tc.currentDiscState;
            ref ForcesState   dForces = ref tc.currentDiscState.forcesState;
            ref DiscModelData dObject = ref tc.discObject;

            // Moments of inertia for a thin disc
            float Ix = 0.25f * dObject.mass * dObject.radius * dObject.radius;
            float Iy = 0.25f * dObject.mass * dObject.radius * dObject.radius;
            float Iz = 0.50f * dObject.mass * dObject.radius * dObject.radius;

            dForces.gyroTorqueY = 0f;
            dForces.gyroTorqueX = 0f;

            // Only compute precession when the disc has meaningful spin
            if (abs(dState.discRotationVel) > 0.1f)
            {
                // Net pitch and roll moments (aero + collision + friction)
                float newPitchMoment = dForces.liftInducedPitchingMoment
                                     + dForces.aeroTorqueY
                                     + dForces.collisionTorqueXYZ.y
                                     + tc.frictionInput.angTorqueXYZ.y;

                float newRollMoment  = 0f
                                     + dForces.aeroTorqueX
                                     + dForces.collisionTorqueXYZ.x
                                     + tc.frictionInput.angTorqueXYZ.x;

                // Gyroscopic precession rates
                // Wp = precession rate about pitch axis (rad/s)
                // Wq = precession rate about roll  axis (rad/s)
                float Wp = -(newRollMoment)  / (Iz * dState.discRotationVel);
                float Wq =  (newPitchMoment) / (Iz * dState.discRotationVel);

                // Angular accelerations needed to drive current vel → precession vel
                float dtSafe = max(dt, CLOSE_TO_ZERO);
                float WqD = (Wq - dState.discRollingVel)  / dtSafe;
                float WpD = (Wp - dState.discPitchingVel) / dtSafe;

                // Convert to torques
                dForces.gyroTorqueY = WpD * Iy;
                dForces.gyroTorqueX = WqD * Ix;
            }
        }
    }
}
